[LOOP-349] Refactor to support service plugins

darinkrauss opened the pull request.

darinkrauss committed:
Refactor to support service plugins

nhamming commented on Loop/View Controllers/SettingsTableViewController.swift:
```diff
@@ -293,8 +293,8 @@ final class SettingsTableViewController: UITableViewController {
-            if indexPath.row < servicesSorted.count {
-                let service = servicesSorted[indexPath.row]
+            if indexPath.row < activeServices.count {
+                let service = activeServicesSorted[indexPath.row]
```
Should the sorted version be used in both instances?

darinkrauss replied:
In the two cases where `activeServices` is used, it actually doesn't matter, but for clarity, yes, I will update.

nhamming commented on Loop/Extensions/UIAlertController.swift:
```diff
@@ -125,25 +125,25 @@ extension UIAlertController {
+    ///   - services: An array of services.
+    ///   - selectionHandler: A closure to execute when a service is selected.
+    ///   - identifier: The identifier of the selected service.
+    internal convenience init(services: [AvailableDevice], selectionHandler: @escaping (_ identifier: String) -> Void) {
```
nit: Could just be me, but I find `services: [AvailableDevice]` confusing and I questioned if this was correct. Maybe a `Typealias` to define `AvailableService`?

darinkrauss replied:
👍

nhamming commented on Loop/Managers/ServicesManager.swift:
```diff
@@ -7,53 +7,114 @@
-            dispatchPrecondition(condition: .onQueue(.main))
-            UserDefaults.appGroup?.services = services
-            notifyObservers()
+    private let observersLock = UnfairLock()
```
Interesting that the observers are also locked. Should this be common practice? I have noticed other places (e.g., DexcomCGMKit) where this isn't the case.

darinkrauss replied:
I'd think so.

nhamming commented on Loop/View Controllers/SettingsTableViewController.swift:
```diff
@@ -721,35 +718,47 @@ extension SettingsTableViewController: CGMManagerSetupViewControllerDelegate {
-        return serviceTypes.filter { serviceType in !dataManager.servicesManager.services.contains { type(of: $0) == serviceType } }
+    fileprivate var availableServices: [AvailableDevice] {
+        return dataManager.servicesManager.availableServices
+    }
```
Similar nit here about `availableServices` defined as `availableDevice`

nhamming reviewed (approved):
Just general comments/questions.

darinkrauss committed:
Updates from PR review

- Add type alias from AvailableDevice to AvailableService for clarity
- Always use sorted active services list in SettingsTableViewController

nhamming reviewed (approved):
LGTM

ps2 commented on Loop.xcodeproj/project.pbxproj:
```diff
@@ -71,7 +71,6 @@
 		4345E40421F68AD9009E00E5 /* TextRowController.swift in Sources */ = {isa = PBXBuildFile; fileRef = 4345E40321F68AD9009E00E5 /* TextRowController.swift */; };
 		4345E40621F68E18009E00E5 /* CarbEntryListController.swift in Sources */ = {isa = PBXBuildFile; fileRef = 4345E40521F68E18009E00E5 /* CarbEntryListController.swift */; };
 		4346D1E71C77F5FE00ABAFE3 /* ChartTableViewCell.swift in Sources */ = {isa = PBXBuildFile; fileRef = 4346D1E61C77F5FE00ABAFE3 /* ChartTableViewCell.swift */; };
-		434FB6461D68F1CD007B9C70 /* Amplitude.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = 434FB6451D68F1CD007B9C70 /* Amplitude.framework */; };
```
Love to see all these files drop out of Loop!

ps2 commented on Loop/AppDelegate.swift:
```diff
@@ -17,13 +17,15 @@ final class AppDelegate: UIResponder, UIApplicationDelegate {
+
+    private lazy var servicesManager = ServicesManager(pluginManager: pluginManager)
 
     private lazy var analyticsServicesManager = AnalyticsServicesManager(servicesManager: servicesManager)
```
This seems inverted. If the ServicesManager owns the other managers, then you can use an imperative style of programming that should save on locks, and replace some of the observe/notify style interactions that are on the ServicesManager with direct calls. I think at this layer of Loop, direct control is probably more appropriate than loosely coupling things.

darinkrauss replied:
I believe I have address your concerns here in the latest go around. Please review and let me know what you think.

darinkrauss committed:
Refactor Service and related delegates from PR review

- Fix service identifier key in RawStateValue

darinkrauss committed:
Rename ServiceDelegate function from PR review

ps2 commented on Loop/Plugins/LoopPlugins.swift:
```diff
@@ -102,13 +104,51 @@ class PluginManager {
         return
             object(forInfoDictionaryKey: LoopPluginBundleKey.pumpManagerIdentifier.rawValue) as? String != nil ||
-            object(forInfoDictionaryKey: LoopPluginBundleKey.cgmManagerIdentifier.rawValue) as? String != nil
+            object(forInfoDictionaryKey: LoopPluginBundleKey.cgmManagerIdentifier.rawValue) as? String != nil ||
```
I think having to open up *all* the framework's plist files is going to be felt as slow by users. My original idea was to name the plugins with a .loopplugin extension, but that doesn't get past app store rules.  I'm now thinking plugin bundles should either contain a special file that denotes them as plugins, or the plugin installer script generates a manifest.  Leaning toward the second idea. No changes needed here; just thinking out loud. :)

darinkrauss replied:
:+1: for the second idea.

ps2 reviewed (commented):
I would consider restructuring the relationship between `ServicesManager` and the subtype manager instances, to simplify the interactions.  I'm not sure the observer pattern (and its associated locking) is needed. If ServicesManager owns and is responsible for the subtypes, they probably don't need to know how they're being managed, and can be controlled directly.

nhamming reviewed (approved):
+1

darinkrauss committed:
Refactor service managers for PR review

- Merge all *ServiceManagers into ServicesManager
- Remove unnecessary ServicesManager support for *ServiceManagers
- Refactor ServiceManager creation and ownership

ps2 reviewed (approved):
LGTM

darinkrauss committed:
Updates from PR review

- Refactor various service managers into separate classes
- Embed various service managers into ServicesManager

darinkrauss commented:
@ps2 Last commit breaks apart the service manager again per your PR review thoughts.

ps2 reviewed (approved):
LGTM!

darinkrauss committed:
Updates based upon PR review
